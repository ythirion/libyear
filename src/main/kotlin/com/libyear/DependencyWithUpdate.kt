package com.libyear

import kotlinx.serialization.Serializable

@Serializable
data class DependencyWithUpdate(
    val name: String,
    val projectVersion: String,
    val latestVersion: String,
    val versionDiff: VersionDiff
) {
    override fun toString(): String =
        "${this.name} ${this.projectVersion} -> ${this.latestVersion} (${this.versionDiff})"
}