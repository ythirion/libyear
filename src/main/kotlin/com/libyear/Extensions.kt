package com.libyear

import kotlinx.serialization.encodeToString
import kotlinx.serialization.json.Json
import org.apache.commons.io.FileUtils
import org.apache.maven.model.Dependency
import org.codehaus.mojo.versions.api.ArtifactVersions
import org.codehaus.mojo.versions.api.UpdateScope
import java.io.File

typealias Dependencies = List<DependencyWithUpdate>

fun Dependency?.description() = "${this?.groupId}:${this?.artifactId}"
fun ArtifactVersions?.latest() = this?.getNewestUpdate(currentVersion, UpdateScope.ANY).toString()

fun ArtifactVersions?.diffWithCurrentVersion() =
    VersionDiff(
        this.diffWithCurrentVersion(UpdateScope.ANY),
        this.diffWithCurrentVersion(UpdateScope.MAJOR),
        this.diffWithCurrentVersion(UpdateScope.MINOR),
        this.diffWithCurrentVersion(UpdateScope.INCREMENTAL) + this.diffWithCurrentVersion(UpdateScope.SUBINCREMENTAL),
    )

fun ArtifactVersions?.diffWithCurrentVersion(scope: UpdateScope): Int =
    this?.getAllUpdates(currentVersion, scope)?.size!!

fun Dependencies.aggReport() =
    "> ${sumOf { it.versionDiff.any }} versions behind for $size dependencies" +
            " -> ${sumOf { it.versionDiff.major }} major(s) | ${sumOf { it.versionDiff.minor }} minor(s) | ${sumOf { it.versionDiff.fixes }} fix(es)"

fun Dependencies.toReport() = Json.encodeToString(this)

fun Dependencies.logResult(log: (info: String) -> Unit): Dependencies {
    when {
        this.isEmpty() -> log("No dependency updates found. WELL DONE !!!")
        else -> {
            log("The following dependencies have newer versions:")
            forEach { log(it.toString()) }
            log(aggReport())
        }
    }
    return this
}

//region File management
fun File.create(): File {
    FileUtils.cleanDirectory(this)
    FileUtils.forceMkdir(this)
    return this
}

fun File?.saveToFile(fileName: String, content: String) {
    FileUtils.writeStringToFile(
        File(this, fileName),
        content,
        Charsets.UTF_8
    )
}
//endregion