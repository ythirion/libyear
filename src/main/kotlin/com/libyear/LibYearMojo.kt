package com.libyear

import org.apache.maven.plugins.annotations.Mojo
import org.apache.maven.plugins.annotations.Parameter
import org.codehaus.mojo.versions.AbstractVersionsUpdaterMojo
import org.codehaus.mojo.versions.rewriting.ModifiedPomXMLEventReader
import java.io.File

@Mojo(name = "run")
class LibYearMojo : AbstractVersionsUpdaterMojo() {
    @Parameter(defaultValue = "\${project.basedir}/target/libyear", readonly = true)
    private val outputDirectory: File? = null

    override fun execute() {
        retrieveDependenciesWithUpdates()
            .logResult { log.info(it) }
            .toReport()
            .runCatching {
                outputDirectory?.create()
                    .saveToFile("${project.name}.json", this)
            }.onFailure { ex -> log.error(ex) }
    }

    private fun retrieveDependenciesWithUpdates(): Dependencies {
        return helper.lookupDependenciesUpdates(project.dependencies.toMutableSet(), false)
            .mapNotNull { (artifact, versions) ->
                DependencyWithUpdate(
                    artifact.description(),
                    artifact.version,
                    versions.latest(),
                    versions.diffWithCurrentVersion()
                )
            }
            .filter { it.versionDiff.any > 0 }
    }

    override fun update(pom: ModifiedPomXMLEventReader?) {
        TODO("Not yet implemented")
    }
}