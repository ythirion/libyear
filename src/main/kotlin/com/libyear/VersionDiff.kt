package com.libyear

import kotlinx.serialization.Serializable

@Serializable
data class VersionDiff(val any: Int, val major: Int, val minor: Int, val fixes: Int) {
    override fun toString(): String {
        return "$any version(s) behind -> $major major(s) | $minor minor(s) | $fixes fix(es)"
    }
}