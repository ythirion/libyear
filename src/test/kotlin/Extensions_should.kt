import com.libyear.DependencyWithUpdate
import com.libyear.VersionDiff
import com.libyear.logResult
import com.libyear.toReport
import org.approvaltests.Approvals
import org.junit.jupiter.api.Test

class Extensions_should {
    private val dependenciesWithUpdates = listOf(
        DependencyWithUpdate("com.libyear",
            "0.0.1",
            "1.0.0",
            VersionDiff(31, 22, 5,3)),
        DependencyWithUpdate("kotlinx.serialization",
            "0.0.1",
            "1.0.0",
            VersionDiff(10, 2, 5,3)))

    @Test
    fun export_dependency_report() {
        Approvals.verify(dependenciesWithUpdates.toReport())
    }

    @Test
    fun log_dependencies() {
        val log = StringBuilder()
        dependenciesWithUpdates.logResult{ log.appendLine(it) }

        Approvals.verify(log.toString())
    }
}