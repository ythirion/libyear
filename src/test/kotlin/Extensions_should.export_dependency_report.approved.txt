[{"name":"com.libyear","projectVersion":"0.0.1","latestVersion":"1.0.0","versionDiff":{"any":31,"major":22,"minor":5,"fixes":3}},{"name":"kotlinx.serialization","projectVersion":"0.0.1","latestVersion":"1.0.0","versionDiff":{"any":10,"major":2,"minor":5,"fixes":3}}]